pytest_container/runtime.py

```diff
@@ -445,34 +435,56 @@ def _get_buildah_version() -> Version:
     )
 
 
+class _AutoDetect:
+    INSTANCE: "_AutoDetect"
+
+
+_AutoDetect.INSTANCE = _AutoDetect()
+
+
 class PodmanRuntime(OciRuntimeBase):
     """The container runtime using :command:`podman` for running containers and
     :command:`buildah` for building containers.
 
     """
 
-    _runtime_functional = LOCALHOST.run("podman ps").succeeded
-    _buildah_functional = LOCALHOST.run("buildah").succeeded
-
-    @staticmethod
-    def _runtime_error_message() -> str:
-        if PodmanRuntime._runtime_functional:
-            return ""
-
-        podman_ps = LOCALHOST.run("podman ps")
-        assert (
-            not podman_ps.succeeded
-        ), "podman runtime is not functional, but 'podman ps' succeeded"
-        return str(podman_ps.stderr)
+    def __init__(
+        self,
+        runner_binary: str = "podman",
+        buildah_binary: Union[str, _AutoDetect, None] = _AutoDetect.INSTANCE,
+    ) -> None:
+        resolved_buildah_binary: Optional[str] = None
+
+        if buildah_binary is _AutoDetect.INSTANCE:
+            # try to guess if argument is omitted
+            buildah_binary = shutil.which("buildah")
+
+        if buildah_binary is None:
+            # if explicitly set to None or not found, we don't have buildah
+            resolved_buildah_binary = None
+            self._buildah_functional = False
+
+        elif isinstance(buildah_binary, str):
+            # make sure it's fully resolved and not just a binary name
+            buildah_binary = shutil.which("buildah")
+            self._buildah_functional = (
+                subprocess.run(
+                    [buildah_binary, "--version"], check=False
+                ).returncode
+                == 0
+                if buildah_binary is not None
+                else False
+            )
+            resolved_buildah_binary = buildah_binary
```

Comment:
Which high level problem are you trying to solve here?

Reply:
As with rest, I don't want the checks to run on module imports and I tried to preserve the original logic of detection if someone creates PodmanRuntime() directly. I can simplify this if you don't really care about the detection in that case and only through get_selected_platform()

Reply:
The detection can be as dead-simple as possible, there's really no need to try be "smart" here as no one requires it (for now).